### PR TITLE
Feature/aql date trunc 2

### PR DESCRIPTION
**Pull request by dothebart**
This superseedes https://github.com/arangodb/arangodb/pull/4742 - which wasn't easily mergeable anymore. http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/444/

### CODE REVIEW

File: arangod/Aql/Functions.cpp [hunk 1]
```diff
@@ -2847,6 +2847,57 @@ AqlValue Functions::DateDaysInMonth(arangodb::aql::Query* query,
...
+  year_month_day ymd{floor<days>(tp)};
+  auto day_time = make_time(tp - sys_days(ymd));
+  milliseconds ms{0};
+
+  if (duration == "y" || duration == "year" || duration == "years") {
+    ymd = year{ymd.year()}/1/1;
+  } else if (duration == "m" || duration == "month" || duration == "months") {
+    ymd = year{ymd.year()}/ymd.month()/1;
+  } else if (duration == "d" || duration == "day" || duration == "days") {
+    ; // ymd = year{ymd.year()}/ymd.month()/ymd.day();
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]: what will happen here?

File: arangod/Aql/Functions.cpp [hunk 2]
```diff
@@ -2847,6 +2847,57 @@ AqlValue Functions::DateDaysInMonth(arangodb::aql::Query* query,
...
+
+  std::string duration = durationType.slice().copyString();
+  std::transform(duration.begin(), duration.end(), duration.begin(), ::tolower);
+
+  year_month_day ymd{floor<days>(tp)};
+  auto day_time = make_time(tp - sys_days(ymd));
+  milliseconds ms{0};
+
+  if (duration == "y" || duration == "year" || duration == "years") {
+    ymd = year{ymd.year()}/1/1;
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]: why divide by 1 two times?

File: arangod/Aql/Functions.cpp [hunk 3]
```diff
@@ -2847,6 +2847,57 @@ AqlValue Functions::DateDaysInMonth(arangodb::aql::Query* query,
...
+  } else if (duration == "m" || duration == "month" || duration == "months") {
+    ymd = year{ymd.year()}/ymd.month()/1;
```
Comment (jsteemann) [inline arangod/Aql/Functions.cpp]: can `ymd.month()` return `0`?

Comment (jsteemann) [reply arangod/Aql/Functions.cpp]: why do we divide by 1 here?

File: arangod/Aql/Functions.cpp [hunk 4]
```diff
@@ -2847,6 +2847,57 @@ AqlValue Functions::DateDaysInMonth(arangodb::aql::Query* query,
   return AqlValue(AqlValueHintUInt(static_cast<uint64_t>(unsigned(lastMonthDay.day()))));
 }
 
+/// @brief function DATE_TRUNC
+AqlValue Functions::DateTrunc(arangodb::aql::Query* query,
+                            transaction::Methods* trx,
+                            VPackFunctionParameters const& parameters) {
+  static char const* AFN = "DATE_TRUNC";
+  using namespace std::chrono;
```
Comment (ObiWahn) [inline arangod/Aql/Functions.cpp]: date and std::chrono implement same types with upcoming standards. We can avoid collisions now!

File: arangod/Aql/Functions.cpp [hunk 5]
```diff
@@ -2847,6 +2847,57 @@ AqlValue Functions::DateDaysInMonth(arangodb::aql::Query* query,
...
+  } else if (duration == "m" || duration == "month" || duration == "months") {
+    ymd = year{ymd.year()}/ymd.month()/1;
+  } else if (duration == "d" || duration == "day" || duration == "days") {
+    ; // ymd = year{ymd.year()}/ymd.month()/ymd.day();
+  } else if (duration == "h" || duration == "hour" || duration == "hours") {
+    ms = day_time.hours();
+  } else if (duration == "i" || duration == "minute" || duration == "minutes") {
+    ms = day_time.hours() + day_time.minutes();
+  } else if (duration == "s" || duration == "second" || duration == "seconds") {
+    ms = day_time.to_duration() - day_time.subseconds();
```
Comment (ObiWahn) [inline arangod/Aql/Functions.cpp]: why subseconds?

File: arangod/Aql/Functions.cpp (hunk 2 above)
Comment (dothebart) [reply arangod/Aql/Functions.cpp]: https://github.com/HowardHinnant/date/wiki/Examples-and-Recipes#obtaining-a-time_point-from-ymd-hms-components demonstrates how the library works with dates and operator overloading.

**Pull request merged**